### DISCUSSION TITLE
OTA update with HTTPS is not working

### DISCUSSION THREAD

**Discussion by wsrrasel** [arendst/Tasmota#10682] [Q&A] [answered]
Hello friends, 
Tasmota supports HTTP with OTA url. But I want to implement HTTPS in Tasmota for my own OTA server. To do this job, I have searched on google and found some solutions. The solutions are for generic ESP8266. I have used the solutions in the tasmota. But It has shown an error. In the case of the generic ESP8266, it worked fine.

**Now, would you please kindly help me that how can I implement HTTPS with OTA url instead of HTTP?**

Here I am adding all the stuff that I have used to overcome my problem
Device: ESP-12E

Added the lines in user_config override.h
![ota_1](https://user-images.githubusercontent.com/11585385/105670125-b96a5f80-5f0a-11eb-8169-da9fcba854a6.PNG)

Added the lines in support_tasmota.ino
![ota_2](https://user-images.githubusercontent.com/11585385/105670201-d4d56a80-5f0a-11eb-9213-ee2bb47a4fc2.PNG)

Error
![ota_3](https://user-images.githubusercontent.com/11585385/105670219-dacb4b80-5f0a-11eb-9ac3-36e785dae7f9.PNG)

The sample code that I have found. It works fine with generic ESP8266
```
#include <ESP8266WiFi.h>
#include <WiFiClientSecure.h>
#include <ESP8266httpUpdate.h>
#include <FS.h>

#define fp "70 94 DE DD E6 C4 69 48 3A 92 70 A1 48 56 78 2D 18 64 E0 B7" //my ota server fingerprint

const char* ssid = "ssid";  // modify this
const char* password = "password";   // modify this

const char* host = "https://myotaserver.com/tasmota/tasmota.bin.gz";

void setup() {
  Serial.begin(115200);
  WiFi.mode(WIFI_STA);
  for (uint8_t t = 4; t > 0; t--)
  {
    Serial.printf("[SETUP] WAIT %d...\n", t);
    Serial.flush();
    delay(1000);
  }
//*/
  Serial.println();
  Serial.print("connecting to ");
  Serial.println(ssid);

  if (WiFi.SSID() != ssid) {
    WiFi.begin(ssid, password);
  }
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println("");
  Serial.println("WiFi connected");
  Serial.println("IP address: ");
  Serial.println(WiFi.localIP());


  WiFiClientSecure client;

  client.setFingerprint(fp);
ESPhttpUpdate.setLedPin(LED_BUILTIN, LOW);
  auto ret = ESPhttpUpdate.update(client, host);
  // if successful, ESP will restart
  Serial.println("update failed");
  Serial.println((int) ret);
}


void loop() {
}
```

**Comment by Jason2866** [comment, marked as answer]
answer from s-hadinger in Discord:

> ESP8266 simply does not have enough memory for the native Arduino HTTPS client implementation. This is way there is a specially optimized TSL client for MQTT. We will not support HTTPS for OTA upgrade, for the reason that tasmota-minimal will never be able to do HTTPS

**Comment by meingraham** [comment]
What you could do is to download the binaries you use to a file server on your own secure network. Then you can point your Tasmota devices to your own web server that has access to those files and can serve those files up via HTTP (but within your secured network).

I'm not sure if HTTPS access to the online binaries could be added... that would be an question for the Tasmota web server admin. That would provide secured downloading of the binaries. But that would still not mean that you would be able to OTA from that secure site due to what has been noted before (limited ESP82xx resources).

**Reply by techman83** [reply]
I personally use NGINX to do this, restricting it to just my IoT subnet:

```conf
server {
	listen          80;
	
	location /tasmota/ {
		allow    192.168.0.0/24;
		deny all;
		alias /var/www/tasmota/;
	}
}
```

**Comment by s-hadinger** [comment]
All the links on ota.tasmota.com also work with https; from your browser not from Tasmota for the reasons stated above.
https://ota.tasmota.com/tasmota/

**Comment by meingraham** [comment]
Purrfect... then, @wsrrasel , you can download the files securely and serve them up to your Tasmota devices on your secure network

**Comment by Jason2866** [comment]
Before doing such a setup i would consider to clone Tasmota github and compile the firmware(s)
with individual settings. Compiling is easy. It just needs python and PlatformIO cli installed.
Just take a look in github actions workflow to see how this is done https://github.com/arendst/Tasmota/blob/development/.github/workflows/Tasmota_build.yml

**Reply by techman83** [reply]
With `platformio_tasmota_cenv.ini` and `user_config_override.h`, it makes having device specific firmware a breeze!

**Reply by ascillato** [reply]
If you don't like using platformio or gitpod, you can use TasmoCompiler ( https://github.com/benzino77/tasmocompiler ) that is an UI to select which features you want and it compiles it.

**Comment by schildbach** [comment]
I wonder if the decision to not provide OTA via HTTPS could be revisited for ESP32? It can do a lot of the crypto in hardware, so I guess at least a certain IoT-friendly subset of TLS could be supported without using too much resources.

Currently, the default of fetching from an HTTP URL is extremely unsafe. A bad actor at some internet exchange could infect millions of Tasmota devices at practically no cost, opening entry points into many private networks.

**Comment by Jason2866** [comment]
@schildbach esp32 does support OTA from https.

**Reply by schildbach** [reply]
Oh sorry, I didn't know. I'm not sure if I changed the OtaUrl, but is it perhaps just `http:` by default? If so, maybe change that to `https:` by default for ESP32?

**Reply by Jason2866** [reply]
It is by default `http:` since `https:` support was added later to the esp32.
Have to check if there is a reason for changing this default.

**Comment by Jason2866** [comment]
@schildbach #17404
